docs/src/development/proposals/MEP9/README.md

```diff
@@ -0,0 +1,111 @@
+
+## Requirements
+
+- Access to firewall SSH only via VPN
+- Easy to update VPN components
+
+As a next step, we can also consider joining the management servers to the VPN mesh, which would replace typical WireGuard setups for operators to enter resources inside the partition.
+
+## High Level Design
+
+[](./architecture.drawio.svg)
+
+> Simplified drawing showing old vs. new architecture.
+
+Store current Headscale and desired Tailscale(`metalctl`?) version in some K8s resource? And allow to update it via `metalctl`.
+
+### Concerns
+
+There's few concerns when using WireGuard for implementing VPN:
+
+1. WireGuard doesn't implement dynamic cipher substitution. Which is important in case one of the crypto methods, used by WireGuard will be broken. The only possible solution for that will be to update WireGuard to a fixed version.
+2. Coordination server(Headscale) is a single point of failure. In case it fails, it potentially can disconnect existing members of the network, as WireGuard can't manage dynamic IPs by itself.
+3. Headscale is already falls behind Tailscale coordination server implementation. Which can complicate the upgrade to newer version of Tailscale client in case of emergency.
+
+### Solutions to concerns
+
+1. Tailscale node software is using userspace implementation of WireGuard -- `wireguard-go`. One of the options is to inject Tailscale client into `metalctl`. And make it available as `metalctl vpn` or similar command. It should be possible to do as `tailscale` node is already available as open sourced Go pkg. That would allow us to control, what version of Tailscale users are using and in case of any critical changes to enforce them to update `metalctl` to use VPN functionality.
+2. Would it be a considerable risk? We could look into `wg-dynamic` project to cover this problem.
+
+## Implementation Details
+
+- Add a headscale server to the metal control plane -- either manually or automate it with `metalctl`? 
+
+### New `metalctl` commands
+
+- `metalctl vpn` -- section for VPN related commands:
+
+    - `metalctl vpn create --name [name] --tag [tag for headscale version] --config [path to config file] --metalctl [minimal metalctl version]` -- will create `headscale` deployment. Do we want to use separate StatefulSet for DB config? If yes, DB credentials have to provided somehow too.
+    - `metalctl vpn update --tag [tag for headscale version] --metalctl [minimal metalctl version]` -- will update `headscale` and/or required minimal `metalctl` version.
+    - `metalctl vpn namespace create [vpn name] --name [namespace name]` -- create new namespace.
+    - `metalctl vpn get key [vpn name] --namespace [namespace name]` -- returns auth key to be used with `tailscale` client for establishing connection.
+    - `metalctl vpn connect [vpn name] --namespace [namespace name]` -- connect to specific namespace. Uses `tailscale` pkg to set up connection.
```

Comment:
Maybe we should try not to give away the information that we are using VPN and try to make it as transparent for the user as possible. Like `metalctl firewall ssh`.

Reply:
Ok. But to be clear, VPN endpoints still will be neaded for metal-api(at least to get auth key).

Reply:
Yep!

Reply:
I dont think metal-api, need a vpn connection, only a grpc connection to the headscale api endpoint

Reply:
Yeah, i mean VPN endpoint, not a connection) So basically an intermediary between client-side and headscale(so client-side doesn't have direct connection to headscale).